Create rule S6460: Template should not be constrained with ad-hoc requires-expression (#1270)
SonarSource/rspec: rules/S6460/cfamily/rule.adoc

Since {cpp}20, it is possible to add a _requires-clause_ to a template as a way to express requirements (constraints) on the template arguments. 
This construct is versatile and allows any expression that evaluates to either `true` or `false` at compile time to be used. 

> _Requires-expressions_ are compile-times expressions that can be used to express required operations on types:

Reviewer: ```suggestion
One of such compile-time expressions are _requires-expressions_ that can be used to express required operations on types:
```

> One of these expressions is the _Requires-expression_, which can be used to express required operations on types:

Reviewer 2: ```suggestion
One of these expressions is the _requires-expression_, which can be used to express required operations on types:
```

[source,cpp]
----
template<typename M>
  requires requires(M a, M b) {
    { a + b };
    { a * b };
  }
M fma(M a, M b, M c) {
  return a + (b * c);
}
----

While the use of an ad-hoc _requires-expression_ provides a way to quickly constrain a template, it limits the readability of the code:
While it conveys the required syntax requirements, it fails to express the semantics associated with the operations. 

> For instance, the above template expects that `+` / `*` perform mathematical additions and multiplications, not a concatenation of strings.

Reviewer 2: ```suggestion
For instance, the above template expects that `+` and `*` perform mathematical additions and multiplications, not a concatenation of strings.
```
(too many consecutive operator characters)

----

This rule raises an issue for any use of an ad-hoc _requires-expression_ in the requirements of template functions, classes, or variables.

== Noncompliant Code Example

[source,cpp]
----
template<typename M>
  requires requires(M a, M b) { // noncompliant
    { a + b };
    { a * b };
  }
M fma(M a, M b, M c) {
  return a + (b * c);
}

template<typename ForwardIt>
  requires requires(ForwardIt it) { // noncompliant
    /* dereference and others */
    ++it; 
  }
ForwardIt rotate(ForwardIt first, ForwardIt mid, ForwardIt last);

template<typename BidirectionalIt>
  requires requires(BidirectionalIt it) { // noncompliant
    /* dereference and others */
    ++it; 
    --it;
  } 
BidirectionalIt rotate(BidirectionalIt first, BidirectionalIt mid, BidirectionalIt last);
----

== Compliant Solution

[source,cpp]
----
template<typename T>
concept Multiplicative = requires(const T a, const T b) {
  { a + b };
  { a * b };
}

template<Multiplicative M>
M fma(M a, M b, M c) {
  return a + (b * c);
}

template<class ForwardIt>
  requires std::forward_iterator<ForwardIt>
ForwardIt rotate(ForwardIt first, ForwardIt mid, ForwardIt last);

> // std::bidirectional_iterator subusumes std::forward_iterator, as it is defined as:

Reviewer 2: ```suggestion
// std::bidirectional_iterator subsumes std::forward_iterator, as it is defined as:
```
